### PR TITLE
New spec validation: Subscriptions root field must not contain @skip nor @include on root selection set

### LINKED ISSUE #3868: Tracking spec change: Prevent @skip and @include on root subscription selection set
This has reached RFC2 stage and it's already in graphql-js, so it's probably getting merged soon

Spec link: https://github.com/graphql/graphql-spec/pull/860

### PR DESCRIPTION
**Pull request by dondonz**
Fixes #3868 and implements https://github.com/graphql/graphql-spec/pull/860

We had an existing validation for subscription root fields. This adds on a new requirement that subscription root fields can't contain the `@include` nor `@skip` directives. See the spec link for more.

This could be considered a "breaking change" but it is more correctly a new validation requirement.

### CODE REVIEW

File: src/main/java/graphql/validation/Validator.java [hunk 1]
```diff
@@ -27,7 +27,7 @@
 import graphql.validation.rules.PossibleFragmentSpreads;
 import graphql.validation.rules.ProvidedNonNullArguments;
 import graphql.validation.rules.ScalarLeaves;
-import graphql.validation.rules.SubscriptionUniqueRootField;
+import graphql.validation.rules.SubscriptionRootField;
```
Comment (dondonz) [inline src/main/java/graphql/validation/Validator.java:30]: I changed this rule name to be more generic, reflecting that it does 2 checks on the root field

File: src/main/java/graphql/validation/rules/SubscriptionRootField.java [hunk 2]
```diff
@@ -39,16 +47,24 @@ public void checkOperationDefinition(OperationDefinition operationDef) {
 
             GraphQLObjectType subscriptionType = getValidationContext().getSchema().getSubscriptionType();
 
+            // This coercion takes into account default values for variables
```
Comment (dondonz) [inline src/main/java/graphql/validation/rules/SubscriptionRootField.java:50]: This is fixing an edge case - this check previously didn't take into account default variable values because there was previously no coercion.

That caused my tests to fail (see later) because I had set a default boolean value for include and skip.

File: src/main/java/graphql/validation/rules/SubscriptionRootField.java [hunk 3]
```diff
@@ -39,16 +47,24 @@ public void checkOperationDefinition(OperationDefinition operationDef) {
...
                     .schema(getValidationContext().getSchema())
                     .fragments(NodeUtil.getFragmentsByName(getValidationContext().getDocument()))
-                    .variables(CoercedVariables.emptyVariables().toMap())
+                    .variables(coercedVariableValues.toMap())
                     .objectType(subscriptionType)
                     .graphQLContext(getValidationContext().getGraphQLContext())
                     .build();
 
             MergedSelectionSet fields = fieldCollector.collectFields(collectorParameters, operationDef.getSelectionSet());
-            List<Selection> subscriptionSelections = operationDef.getSelectionSet().getSelections();
```
Comment (dondonz) [inline src/main/java/graphql/validation/rules/SubscriptionRootField.java:51]: Unused variable

File: src/main/java/graphql/validation/rules/SubscriptionRootField.java [hunk 4]
```diff
@@ -57,16 +73,30 @@ public void checkOperationDefinition(OperationDefinition operationDef) {
...
         }
     }
 
     private boolean isIntrospectionField(MergedField field) {
            return field.getName().startsWith("__");
     }
+
+    private boolean hasSkipOrIncludeDirectives(MergedField field) {
+        List<Directive> directives = field.getSingleField().getDirectives();
+        for (Directive directive : directives) {
```
Comment (dondonz) [inline src/main/java/graphql/validation/rules/SubscriptionRootField.java:95]: Given recent performance improvements I am using plain for loops

File: src/main/resources/i18n/Validation.properties [hunk 5]
```diff
@@ -68,9 +68,8 @@ ScalarLeaves.subselectionOnLeaf=Validation error ({0}) : Subselection not allowe
 ScalarLeaves.subselectionRequired=Validation error ({0}) : Subselection required for type ''{1}'' of field ''{2}''
 #
 SubscriptionUniqueRootField.multipleRootFields=Validation error ({0}) : Subscription operation ''{1}'' must have exactly one root field
-SubscriptionUniqueRootField.multipleRootFieldsWithFragment=Validation error ({0}) : Subscription operation ''{1}'' must have exactly one root field with fragments
```
Comment (dondonz) [inline src/main/resources/i18n/Validation.properties:71]: Unused messages

File: src/main/resources/i18n/Validation_de.properties [hunk 6]
```diff
@@ -60,9 +60,9 @@ ScalarLeaves.subselectionOnLeaf=Validierungsfehler ({0}) : Unterauswahl für Bla
 ScalarLeaves.subselectionRequired=Validierungsfehler ({0}) : Unterauswahl erforderlich für Typ ''{1}'' des Feldes ''{2}''
 #
 SubscriptionUniqueRootField.multipleRootFields=Validierungsfehler ({0}) : Subscription operation ''{1}'' muss genau ein root field haben
-SubscriptionUniqueRootField.multipleRootFieldsWithFragment=Validierungsfehler ({0}) : Subscription operation ''{1}'' muss genau ein root field mit Fragmenten haben
 SubscriptionIntrospectionRootField.introspectionRootField=Validierungsfehler ({0}) : Subscription operation ''{1}'' root field ''{2}'' kann kein introspection field sein
-SubscriptionIntrospectionRootField.introspectionRootFieldWithFragment=Validierungsfehler ({0}) : Subscription operation ''{1}'' fragment root field ''{2}'' kann kein introspection field sein
+SubscriptionRootField.forbidSkipAndIncludeOnSubscriptionRoot=Validierungsfehler ({0}) : Subscription operation ''{1}'' root field ''{2}'' darf weder @skip noch @include directive in top level selection
+#
```
Comment (dondonz) [inline src/main/resources/i18n/Validation_de.properties:65]: Check my German please

File: src/test/groovy/graphql/validation/rules/SubscriptionRootFieldNoSkipNoIncludeTest.groovy [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+import graphql.validation.ValidationError
+import graphql.validation.Validator
+import spock.lang.Specification
+
+class SubscriptionRootFieldNoSkipNoIncludeTest extends Specification {
+
+    def "valid subscription with @skip and @include directives on subfields"() {
+        given:
+        def query = """
+        subscription MySubscription(\$bool: Boolean = true) {
```
Comment (dondonz) [inline src/test/groovy/graphql/validation/rules/SubscriptionRootFieldNoSkipNoIncludeTest.groovy:14]: This is what I mean by default values. When an empty coerced variables map is passed, this test fails

File: src/main/resources/i18n/Validation_nl.properties [hunk 8]
```diff
@@ -58,9 +58,8 @@ ScalarLeaves.subselectionOnLeaf=Validatiefout ({0}) : Sub-selectie niet toegesta
 ScalarLeaves.subselectionRequired=Validatiefout ({0}) : Sub-selectie verplicht voor type ''{1}'' van veld ''{2}''
 #
 SubscriptionUniqueRootField.multipleRootFields=Validatiefout ({0}) : Subscription operation ''{1}'' moet exact één root field hebben
-SubscriptionUniqueRootField.multipleRootFieldsWithFragment=Validatiefout ({0}) : Subscription operation ''{1}'' moet exact één root field met fragmenten hebben
 SubscriptionIntrospectionRootField.introspectionRootField=Validatiefout ({0}) : Subscription operation ''{1}'' root field ''{2}'' kan geen introspectieveld zijn
-SubscriptionIntrospectionRootField.introspectionRootFieldWithFragment=Validatiefout ({0}) : Subscription operation ''{1}'' fragment root field ''{2}'' kan geen introspectieveld zijn
+SubscriptionRootField.forbidSkipAndIncludeOnSubscriptionRoot=Validation error ({0}) : Subscription operation ''{1}'' root field ''{2}'' must not use @skip nor @include directives in top level selection
```
Comment (dondonz) [inline src/main/resources/i18n/Validation_nl.properties:62]: Add English placeholder

**Review by bbakerman — APPROVED** [review]
